crates/red_knot_python_semantic/src/builtins.rs

```diff
@@ -0,0 +1,12 @@
+use red_knot_module_resolver::resolve_builtins;
+
+use crate::semantic_index::module_global_scope;
+use crate::{semantic_index::symbol::ScopeId, Db};
+
+#[salsa::tracked]
+pub(crate) fn builtins_scope(db: &dyn Db) -> Option<ScopeId<'_>> {
```

Comment:
I wonder if we should use specify in this query to create the builtin module instead of having two queries. Or are there other cases where we need access to the builtins module?

Reply:
That's an interesting question! I don't think there are any cases that come to mind where we should be short-circuiting the normal module-resolution process in the same way, no. So perhaps the new function that we publicly expose from the module resolver should not be Salsa-tracked at all, just a regular function, since we know it will only be called from this one place in the `red_knot_python_semantic` crate, and that once place _will_ be Salsa-tracked.

Reply:
Actually I think what would make sense here for now is simply to make `builtins_module` a regular function instead of a salsa query. No need for specify, but also not reason for `builtins_module` to be a query until/unless we have other callers of it, and so we need separate caching at that layer. For now the caching of `builtins_scope` is sufficient.

I'll make that change.

Reply:
Kind if related. What happens if someone does from builtins import string.

I would suspect that this would fo through the module resolver unless we handle it at every call site (which I rather avoid).

Reply:
> Kind of related. What happens if someone does from builtins import string.
> 
> I would suspect that this would fo through the module resolver unless we handle it at every call site (which I rather avoid).

Yeah, that should just go through normal module resolution machinery. Python won't apply any special-casing there — if there's a first-party module called `builtins`, it will completely shadow the stdlib builtins module for the "explicit" import — so we shouldn't special-case it either. It's only the "implicit" builtins scope that we need to special-case when it comes to module resolution.

Reply:
Except we kind of have to for the case where there is no first party module with that name. Otherwise resolve module returns a different Module instance (which means we reparse and reinfer it)

Reply:
We should add two tests. One where there's no first party builtins and one where there is. For the first case, the module on builtins scope should resolve to the same module as when calling resolve_module

Reply:
Very good points

Reply:
Good call, will add this.

Reply:
I think what's important here is that they are the same `File`, not the same `Module`. `File` is the actual ingredient we use everywhere which has identity semantics; `Module` is just a struct to hold a name and a `File`, which I think we should treat with value semantics (i.e. two different `Module` struct with same name and `File` are equivalent); we discard the `Module` pretty much immediately in type inference and don't use it anywhere.

Imported builtins and implicit builtins via `builtins_scope()` are already the same `File` (but I am adding a test for this).

Reply:
In fact I think we should just change `builtins_module` to `builtins_file` instead; if `builtins_scope` is the only caller, it's silly to wrap the file in a `Module` which we then immediately discard in `builtins_scope`, which only needs the file. Then there is no question of them being "the same Module" because there is no Module at all in the `builtins_scope` path.

Reply:
Regarding first-party code overriding builtins module, this isn't actually possible at runtime unless you mess with importer hooks (which we don't need to -- and can't possibly -- support). If you check `sys.meta_path`, `BuiltinsImporter` comes first, before `PathFinder`, which is the hook that actually implements all the normal filesystem module resolution. So in any normal Python installation, `import sys` and `import builtins` _always_ resolve to the built-in `sys` and `builtins` modules, regardless of first-party code.

This will require some special handling in the module resolver, and I don't think it needs to be done in this PR; I've created https://github.com/astral-sh/ruff/issues/12398 to track it as a separate change.

Reply:
Blegh. I forgot that Python special-cased those modules. Agree with deferring those to a followup.

Reply:
Considering that the module resolver needs special handling for built-ins anyway (#12398,) my preferred solution is to call `resolve_module("builtins")` here instead of resolving the file manually. Yes, there's an extra cost to it but it should be marginal in comparison to checking an entire project and any later `resolve_module("builtins")` would be free after that (or as free as Salsa makes a query lookup).

Reply:
I think we should leave it as-is for now. We'll need a routine that's _like_ this for short-circuiting "normal module resolution" when we encounter a builtin module anyway. So I think #12398 will be easier to implement if we keep this as a separate routine for now. When we fix #12398, we can generalise the routine so that we can short-circuit other special-cased modules as well.

Reply:
> We'll need a routine that's like this for short-circuiting "normal module resolution" when we encounter a builtin module anyway.

That's correct. What I understand is that this routine would live in `resolve_module`.

Reply:
Looking at this again, I think I agree with Micha -- it seems like the simplest implementation of this "special case" ultimately is just a bit of added code in `resolve_name` to skip non-stdlib search paths if the name being resolved is in the builtin-modules set. Once we have that, `builtins_scope` may as well just `resolve_module("builtins")`. So I don't see where a separate routine is actually really needed.

Also it looks to me like the fix for #12398 is likely to be so small (literally it looks like just adding the set of builtin module names, then one `if not search_path.is_standard_library() and builtin_module_names.contains(name) { continue; }` in `resolve_name`?) that maybe we may as well just do it in this PR?

Reply:
> Looking at this again, I think I agree with Micha -- it seems like the simplest implementation of this "special case" ultimately is just a bit of added code in `resolve_name` to skip non-stdlib search paths if the name being resolved is in the builtin-modules set. Once we have that, `builtins_scope` may as well just `resolve_module("builtins")`. So I don't see where a separate routine is actually really needed.

Yes, fair enough

> Also it looks to me like the fix for #12398 is likely to be so small (literally it looks like just adding the set of builtin module names, then one `if not search_path.is_standard_library() and builtin_module_names.contains(name) { continue; }` in `resolve_name`?) that maybe we may as well just do it in this PR?

If we just hardcode the list of builtin names into `resolver.rs`, yes. Longer term I think we probably want to have generated Rust code like in https://github.com/astral-sh/ruff/blob/main/crates/ruff_python_stdlib/src/sys.rs, which will require writing a Python script like in https://github.com/astral-sh/ruff/blob/main/scripts/generate_known_standard_library.py. Which also won't be that much work, but is big enough to warrant its own PR, I think. But I'm fine with just hardcoding a list in `resolver.rs` for now.